Data.Rmd

```diff
@@ -0,0 +1,462 @@
+#' @export
+mfl <- function() {
+  favorite_letters
+}
+
+#' Change my favorite letters
+#' @export
+set_mfl <- function(l = letters[24:26]) {
+  favorite_letters <<- l
+  favorite_letters
+}
+```
+
+`favorite_letters` is initialized to ("a", "b", "c") when the package is built.
+The user can then inspect `favorite_letters` with `mfl()`, at which point they'll probably want to register *their* favorite letters with `set_mfl()`.
+Note that we've used the super assignment operator `<<-` in `set_mfl()` in the hope that this will reach up into the package environment and modify the internal data object `favorite_letters`.
+But a call to `set_mfl()` fails like so:
+
+```{r, eval = FALSE}
+set_mfl(c("j", "f", "b"))
+#> Error in set_mfl(c("j", "f", "b")): cannot change value of locked binding for
+#> 'favorite_letters'
+```
+
+Because `favorite_letters` is a regular character vector, modification requires making a copy and rebinding the name `favorite_letters` to this new value.
+And that is what's disallowed: you can't change the binding for objects in the package namespace (well, at least not without trying harder than this).
+
+However, if we maintain our internal state within an internal package environment, we can modify objects contained in the environment and even add completely new objects.
+Here's an alternative implementation that uses an internal environment named `the`.
+
+```{r, eval = FALSE}
+the <- new.env(parent = emptyenv())
+the$favorite_letters <- letters[1:3]
+
+#' Report my favorite letters
+#' @export
+mfl2 <- function() {
+  the$favorite_letters
+}
+
+#' Change my favorite letters
+#' @export
+set_mfl2 <- function(l = letters[24:26]) {
+  the$favorite_letters <- l
+  the$favorite_letters
+}
+```
+
+Now a user *can* record their favorite letters:
+
+```{r, eval = FALSE}
+mfl2()
+#> [1] "a" "b" "c"
+
+set_mfl2(c("j", "f", "b"))
+#> [1] "j" "f" "b"
+
+mfl2()
+#> [1] "j" "f" "b"
+```
+
+We often like to use the name "the" for an internal package environment, because it allows you to refer to the objects inside in a very natural way, such as `the$token`, meaning "the token".
+It is also important to specify `parent = emptyenv()` for such an internal environment, as you generally don't want the environment to inherit from any other (nonempty) environment.
+
+Here are some examples of how packages use this technique:
+
+-   Various functions in the googledrive package need to know the file ID for the current user's home directory on Google Drive. This is an eye-watering string of 40 seemingly random characters that means nothing to a human. It would be inhumane to expect a user to know this or to pass it into every function. Instead, googledrive determines this ID upon first need, then caches it for later use.
+-   TODO: Need 1-2 more examples.
+
+The blog post [Package-Wide Variables/Cache in R Packages](https://trestletech.com/2013/04/package-wide-variablescache-in-r-package/) gives a more detailed development of this technique.
```

Comment:
Worth noting that `load_all()` will wipe out the every time.

I wonder if it's worth comparing `the` to `options()`?

Reply:
I haven't added this yet, because the section feels like it's already getting very long. But I'll keep it mind as we continue to iterate on this.

Reply:
Correction: I did add the info about `load_all()` but not the comparison to `options()`.